### PR TITLE
Paper: Improving Code Quality with Array and DataFrame Type Hints

### PR DESCRIPTION
**Pull request by flexatone**
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_: Charles Lindsey @cdlindsey

_Reviewers_: 
- Victoria Adesoba @vicadesoba 
- Talha Irfan @EngineerKhan

### CODE REVIEW

**Comment by EngineerKhan** [comment]
I have read this paper thrice and cannot find any issue with it. It's well-written, so easy to follow and builds a narrative in an excellent manner.

**Some Minor Suggestions** (can be ignored):

- "_Just like a dictionary, a DataFrame is a complex data structure composed of many component types: the type of the index labels, the type of the column labels, and the types of column values._" - the type of part is a bit redundant and can be rewritten.
- Raised errors are there in the code comments, but some of them (as per the author's discretion) may be explained a bit too.
- "_The same annotations can be used for runtime validation. While reliance on duck-typing over runtime validation is common in Python, runtime validation is often needed with complex data structures such as arrays and DataFrames._" - maybe we can give an example to support the claim?
- At some point, the section under "DataFrame Type Annotations" becomes a bit hard to read. Maybe we can add some subheadings there? Again, this is something the author can better determine.

Overall, a well-written article. Not only, I have learnt new things, but also enjoyed reading it with its smooth style.

**Comment by flexatone** [comment]
Many thanks, @EngineerKhan , for your suggestions and feedback.

**Comment by vicadesoba** [comment]
@flexatone Glad to review your paper and nice to meet you! Just a quick intro: I got introduced to the open source community during my time at makepath a geospatial data science company that maintains several open source packages and I'm involved with Project Bokeh (as much as time allows). Recently, I started working for a process engineering company that isn't involved in the open source community so volunteering with SciPy gives me a small way to stay involved!

File: papers/christopher_ariza/main.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+TSeriesYMBool = sf.Series[sf.IndexYearMonth, np.bool_]
+TSeriesDFloat = sf.Series[sf.IndexDate, np.float64]
+
+def process5(v: TFrameDateInts, q: TSeriesYMBool) -> TSeriesDFloat:
+    t = v.index.iter_label().apply(lambda l: q[l.astype('datetime64[M]')]) # type: ignore
+    s = np.where(t, 0.5, 0.25)
+    return cast(TSeriesDFloat, (v.via_T * s).mean(axis=1))
+```
+
+As before, ``mypy`` can be used to validate annotated interfaces. Below, a ``Frame`` without the expected column value types is passed, causing ``mypy`` to error (shown as comments, below).
```
Comment (vicadesoba) [inline papers/christopher_ariza/main.md]: Just a minor comment, as it doesn't affect readability that much, but can you think of a suitable replacement for "As before"? What specifically is "before" referring too?

File: papers/christopher_ariza/main.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+x = process6(v6, q) # a Frame with three integer columns passes
+```
+As with NumPy arrays, ``Frame`` annotations can wrap ``Require`` specifications in ``Annotated`` generics, permitting the definition of additional run-time validations.
+
+## Conclusion
+
+Python type annotations can make static analysis of types a valuable check of code quality, discovering errors before code is even executed. Up until recently, an interface might take an array or a DataFrame, but no specification of the types contained in those containers was possible. Now, complete specification of component types is possible in NumPy and StaticFrame, permitting more powerful static analysis of types.
+
+Providing correct type annotations is an investment. Reusing those annotations for runtime checks provides the best of both worlds. StaticFrame's ``CallGuard`` runtime type checker is specialized to correctly evaluate fully specified generic NumPy types, as well as all generic StaticFrame containers.
+
```
Comment (vicadesoba) [inline papers/christopher_ariza/main.md:282]: Your continuous use of examples to substantiate your points made your paper easy to read. And you tied together your "thesis" well in the conclusion. Nicely done!

Looking forward to following along as you fine-tine what is already a well-thought-out paper!

File: papers/christopher_ariza/main.md (hunk 1 above)
Comment (flexatone) [reply papers/christopher_ariza/main.md]: Good point! Instead, I have the following:
> These more complex annotated interfaces can also be validated with ``mypy``.

File: papers/christopher_ariza/main.md (hunk 2 above)
Comment (flexatone) [reply papers/christopher_ariza/main.md:282]: Many thanks for your comments and feedback.

**Comment by cdlindsey** [comment]
Thanks for the interesting paper.  I think it's worth mentioning how these issues are handled (or not) with TensorFlow and Pandas.

**Comment by flexatone** [comment]
Many thanks, @cdlindsey , for your feedback. I have added a small section describing how these issues are handled in PyTorch, TensorFlow, and Pandas.

**Review by EngineerKhan — APPROVED** [review]
